src/offset.rs

```diff
@@ -490,6 +491,108 @@ where
     }
 }
 
+/// An owned iterator over an offset
+pub struct OffsetIntoIter<T, OffsetItem: OffsetElement, Buffer: BufferType>
+where
+    T: IntoIterator,
+    <Buffer as BufferType>::Buffer<OffsetItem>: IntoIterator,
+{
+    /// The underlying array data
+    data: <T as IntoIterator>::IntoIter,
+    /// A peekable offsets buffer
+    offsets: Peekable<<<Buffer as BufferType>::Buffer<OffsetItem> as IntoIterator>::IntoIter>,
+}
+
+impl<T, OffsetItem: OffsetElement, Buffer: BufferType> Iterator
+    for OffsetIntoIter<T, OffsetItem, Buffer>
+where
+    T: IntoIterator,
+    <Buffer as BufferType>::Buffer<OffsetItem>: IntoIterator<Item = OffsetItem>,
+{
+    type Item = Vec<<T as IntoIterator>::Item>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let (Some(start), Some(&end)) = (self.offsets.next(), self.offsets.peek()) {
+            let slot_start: usize = start.try_into().expect("offset value should be in range");
+            let slot_end: usize = end.try_into().expect("offset value should be in range");
+
+            assert!(
+                slot_end >= slot_start,
+                "offsets must be monotonically increasing, but a slot's start \
+                offset value {slot_start} is greater than the slot's end offset \
+                value {slot_end}, resulting in a negative slot length"
+            );
+
+            let slot_length = slot_end - slot_start;
+            let taken = self.data.by_ref().take(slot_length).collect::<Vec<_>>();
+
+            assert!(
```

Comment:
We can assume correct inputs in these impls:
```suggestion
            debug_assert!(
```